lib/routes/_core_profile.js

```diff
@@ -0,0 +1,91 @@
+const config = require('../config');
+const logger = require('../logging')('routes._core_profile');
+const request = require('../request');
+
+const AUTH_SERVER_URL = config.get('authServer.url') + '/account/profile';
+
+/**
+ * This is an internal-use route that retreives all the user
+ * profile data available from the auth-server.  Various public
+ * routes can call into it and return a subset of the info.
+ */
+module.exports = {
+  isInternal: true,
+  auth: {
+    strategy: 'oauth',
+    scope: ['profile', 'profile:email', 'profile:locale', 'profile:amr', /* openid-connect scope */'email' ]
+  },
+  response: {
+    schema: {
+      email: Joi.string().optional(),
+      locale: Joi.string().optional(),
+      amrValues: Joi.array().items(Joi.string().required()).optional(),
+      twoFactorAuthentication: Joi.boolean().optional()
+    }
+  },
+  handler: function _core_profile(req, reply) {
+    request.get(AUTH_SERVER_URL, {
+      headers: {
+        Authorization: 'Bearer ' + req.auth.credentials.token
+      },
+      json: true
+    }, (err, res, body) => {
+      if (err) {
+        logger.error('request.auth_server.network', err);
+        return reply(new AppError.authError('network error'));
+      }
+      if (res.statusCode >= 400) {
+        body = body && body.code ? body : { code: res.statusCode };
+        if (res.statusCode >= 500) {
+          logger.error('request.auth_server.fail', body);
+          return reply(new AppError.authError('auth-server server error'));
+        }
+        // Return Unauthorized if the token turned out to be invalid,
+        // or if the account has been deleted on the auth-server.
+        // (we can still have valid oauth tokens for deleted accounts,
+        // because distributed state).
+        if (body.code === 401 || body.errno === 102) {
+          logger.info('request.auth_server.fail', body);
+          return reply(new AppError.unauthorized(body.message));
+        }
+        // There should be no other 400-level errors, unless we're
+        // sending a badly-formed request of our own.  That warrants
+        // an "Internal Server Error" on our part.
+        logger.error('request.auth_server.fail', body);
+        return reply(new AppError({
+          code: 500,
+          message: 'error communicating with auth server'
+        }));
+      }
+
+      if (! body) {
+        return reply(
+          new AppError('empty body from auth response')
+        );
+      }
+      const result = {};
+      if (typeof body.email !== 'undefined') {
+        result.email = body.email;
+      }
+      if (typeof body.locale !== 'undefined') {
+        result.locale = body.locale;
+      }
+      // Translate from internal terminology into OAuth-style terminology.
+      if (typeof body.authenticationMethods !== 'undefined') {
+        result.amrValues = body.authenticationMethods;
+      }
+      if (typeof body.authenticatorAssuranceLevel !== 'undefined') {
```

Comment:
If no `authenticatorAssuranceLevel` set, any benefit to setting `result.twoFactorAuthentication` to false?

Reply:
I thought about it a little, but I kinda like the symmetry of not reporting anything if the auth-server doesn't report anything.  This way if we happen to disable the AAL stuff over in auth-server due to e.g. a bug, the profile-server will end up saying "I don't know" rather than "no 2fa here".